core/arch/riscv64/emit_utils.c

```diff
@@ -558,32 +612,99 @@ append_call_exit_dr_hook(dcontext_t *dcontext, instrlist_t *ilist, bool absolute
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    APP(ilist, RESTORE_FROM_DC(dcontext, DR_REG_A0, XFLAGS_OFFSET));
+    APP(ilist,
+        INSTR_CREATE_csrrw(dcontext, opnd_create_reg(DR_REG_X0),
+                           opnd_create_reg(DR_REG_A0),
+                           opnd_create_immed_int(/*fcsr*/ 0x003, OPSZ_12b)));
 }
 
 /* dcontext is in REG_DCXT; other registers can be used as scratch.
  */
 void
 append_restore_simd_reg(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    /* Nothing. */
 }
 
+/* Append instructions to restore gpr on fcache enter, to be executed
+ * right before jump to fcache target.
+ * - dcontext is in REG_DCXT
+ * - DR's tls base is in dr_reg_stolen
+ * - all other registers can be used as scratch, and we are using a0.
+ */
 void
 append_restore_gpr(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    APP(ilist, RESTORE_FROM_DC(dcontext, SCRATCH_REG0, REG_OFFSET(dr_reg_stolen)));
+    APP(ilist, SAVE_TO_TLS(dcontext, SCRATCH_REG0, TLS_REG_STOLEN_SLOT));
+
+    APP(ilist, RESTORE_FROM_DC(dcontext, SCRATCH_REG0, REG_OFFSET(DR_REG_TP)));
+    APP(ilist, SAVE_TO_TLS(dcontext, SCRATCH_REG0, TLS_REG_TP_SLOT));
+
+    for (int i = 1; i < 32; i++) {
+        if (i == (REG_DCXT - DR_REG_X0) || i == (DR_REG_TP - DR_REG_X0) ||
+            i == (dr_reg_stolen - DR_REG_X0))
+            continue;
+        APP(ilist, RESTORE_FROM_DC(dcontext, DR_REG_X0 + i, REG_OFFSET(DR_REG_X0 + i)));
+    }
+    APP(ilist, RESTORE_FROM_DC(dcontext, REG_DCXT, REG_OFFSET(REG_DCXT)));
 }
 
+/* Append instructions to save gpr on fcache return, called after
+ * append_fcache_return_prologue.
+ * Assuming the execution comes from an exit stub via jr a1,
+ * dcontext base is held in REG_DCXT, and exit stub in a0.
+ * App's a0 and a1 is stored in TLS_REG0_SLOT and TLS_REG1_SLOT
+ * - store all registers into dcontext's mcontext
+ * - restore REG_DCXT app value from TLS slot to mcontext
+ * - restore dr_reg_stolen app value from TLS slot to mcontext
+ */
 void
 append_save_gpr(dcontext_t *dcontext, instrlist_t *ilist, bool ibl_end, bool absolute,
                 generated_code_t *code, linkstub_t *linkstub, bool coarse_info)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    if (linkstub != NULL) {
+        /* FIXME i#3544: NYI for coarse-grain stub. */
+        ASSERT_NOT_IMPLEMENTED(false);
+    }
+
+    /* a0 and a1 will always have been saved in TLS slots before executing
+     * the code generated here. See, for example:
+     * emit_do_syscall_common, emit_indirect_branch_lookup, handle_sigreturn,
+     * insert_exit_stub_other_flags, execute_handler_from_{cache,dispatch},
+     * transfer_from_sig_handler_to_fcache_return
+     */
+    for (int i = 1; i < 32; i++) {
```

Comment:
An alternative would be

for (int reg = DR_REG_X0 + 1; reg < DR_REG_X0 + 32; reg++) {

and use reg instead of DR_REG_X0 below.

Reply:
Done.